src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs

```diff
@@ -318,22 +486,80 @@ internal void NotifyWorkspaceHosts(Action<IVisualStudioWorkspaceHost> action)
             }
         }
 
+        /// <summary>
+        /// Attempts to get single project by given output binary filePath.
+        /// </summary>
+        /// <remarks>This method may be called on a background thread.</remarks>
         internal bool TryGetProjectByBinPath(string filePath, out AbstractProject project)
         {
-            project = null;
+            lock (_projectsByBinPath)
+            {
+                project = null;
+
+                ImmutableArray<AbstractProject> projects;
+                if (_projectsByBinPath.TryGetValue(filePath, out projects))
+                {
+                    // If for some reason we have more than one referencing project, it's ambiguous so bail
+                    if (projects.Length == 1)
+                    {
+                        project = projects[0];
+                        return true;
+                    }
+                }
 
-            List<AbstractProject> projects;
-            if (_projectsByBinPath.TryGetValue(filePath, out projects))
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Attempts to get the projects by given output binary filePath.
+        /// </summary>
+        /// <remarks>This method may be called on a background thread.</remarks>
+        internal bool TryGetProjectsByBinPath(string filePath, out ImmutableArray<AbstractProject> projects)
+        {
+            lock (_projectsByBinPath)
```

Comment:
Actually, I did add a common `_gate` object but just forgot to switch few lock statements to use it - I'll update the PR.